src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/PackageSourceTelemetry.cs

```diff
@@ -0,0 +1,393 @@
+                    telemetry[PropertyNames.Nupkg.Timing.Max] = data.Nupkg.EventTiming.MaxDuration.TotalMilliseconds;
+                }
+
+                if (data.Nupkg.HeaderTiming != null && data.Nupkg.HeaderTiming.Requests > 0)
+                {
+                    telemetry[PropertyNames.Nupkg.Header.Timing.Min] = data.Nupkg.HeaderTiming.MinDuration.TotalMilliseconds;
+                    telemetry[PropertyNames.Nupkg.Header.Timing.Mean] = data.Nupkg.HeaderTiming.TotalDuration.TotalMilliseconds / data.Nupkg.HeaderTiming.Requests;
+                    telemetry[PropertyNames.Nupkg.Header.Timing.Max] = data.Nupkg.HeaderTiming.MaxDuration.TotalMilliseconds;
+                }
+            }
+
+            return telemetry;
+        }
+
+        private static TelemetryEvent ToStatusCodeTelemetry(Dictionary<int, int> statusCodes)
+        {
+            var subevent = new TelemetryEvent(null);
+
+            foreach (var pair in statusCodes)
+            {
+                subevent[pair.Key.ToString()] = pair.Value;
+            }
+
+            return subevent;
+        }
+
+        private static string GetMsFeed(PackageSource source)
+        {
+            if (source.IsHttp)
+            {
+                if (TelemetryUtility.IsNuGetOrg(source))
+                {
+                    return "nuget.org";
+                }
+                else if (TelemetryUtility.IsAzureArtifacts(source))
+                {
+                    return "Azure Artifacts";
+                }
+                else if (TelemetryUtility.IsGitHub(source))
+                {
+                    return "GitHub Package Registry";
+                }
+            }
+            else if (source.IsLocal)
+            {
+                if (TelemetryUtility.IsVsOfflineFeed(source))
+                {
+                    return "VS Offline";
+                }
+            }
+
+            return null;
+        }
+
+        internal class Data
+        {
+            internal ResourceData Metadata { get; }
+            internal ResourceData Nupkg { get; }
+
+            internal Data()
+            {
+                Metadata = new ResourceData();
+                Nupkg = new ResourceData();
+            }
+        }
+
+        internal class ResourceData
+        {
+            public object Lock = new object();
+            public ResourceTimingData EventTiming = new ResourceTimingData();
+            public ResourceTimingData HeaderTiming;
+            public long TotalBytes;
+            public long MaxBytes;
+            public Dictionary<int, int> StatusCodes = new Dictionary<int, int>();
+            public int Successful;
+            public int Retries;
+            public int Cancelled;
+            public int Failed;
+        }
+
+        internal class ResourceTimingData
+        {
+            public int Requests;
+            public TimeSpan TotalDuration;
+            public TimeSpan MinDuration = TimeSpan.MaxValue;
```

Comment:
I know this default value facilitates [this comparison](https://github.com/NuGet/NuGet.Client/pull/3105/files#diff-ac99d30854ddd78e4f424b9d3550d47fR103), but is it possible that `MaxValue` could be logged in telemetry?  If yes, I'd prefer that some post-processing check changes `MaxValue` to `Zero`.

Reply:
It should not be possible, no.

The value is only used when `Requests` is `> 0`, and `ApplyTiming` will set `MinDuration` in the same call that increments `Requests`.